### ISSUE TITLE
Detect when default value of dataclass field is mutable

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#10553]
The typing spec conformance suite was [recently updated](https://github.com/python/typing/pull/2017) to validate that type checkers are detecting a case that generates a runtime exception when a dataclass field has a default value that is a mutable type. 

```python
@dataclass
class DC5:
    # This should generate an error because a default value of
    # type list, dict, or set generate a runtime error.
    x: list[int] = []
```

Pyright doesn't currently check for this condition.

**Comment by Viicos** [comment]
Hello @erictraut, I think this is going to cause a lot of churn for libraries that accept mutable default values (e.g. Pydantic). I'm not necessarily asking for this to be reverted, but wanted to let you know as this might create an influx of issues.

fwiw, ruff as a [lint rule](https://docs.astral.sh/ruff/rules/mutable-dataclass-default/) for this, and has special cases for libraries such as Pydantic. Of course such special casing shouldn't happen in Pyright, but more generally I think this shows that we may need more fine grained `@dataclass_transform` options (related threads: https://discuss.python.org/t/89531, https://discuss.python.org/t/87963, https://discuss.python.org/t/69067).

**Comment by erictraut** [comment]
Thanks for the input. Yeah, I'm not happy about this change either. I'm not entirely convinced that it should be the job of a type checker to enforce this particular limitation of dataclass. My preference is to eliminate this from the conformance test suite.

I think I'll revert this issue in the meantime and push to modify the [conformance suite](https://github.com/python/typing/blob/9e4ac4732438ecc1231a83ee0e6e0a8990046687/conformance/tests/dataclasses_inheritance.py#L47) instead.

**Comment by erictraut** [comment]
Here's the [PR for the conformance](https://github.com/python/typing/pull/2022) suite update. I don't think this change will be controversial.
